nav2_map_server/src/vo_server/vector_object_server.cpp

```diff
@@ -0,0 +1,582 @@
+
+    if (new_shape) {
+      // Creating new circle
+      try {
+        circle = std::make_shared<Circle>(node, new_params);
+        shapes_.push_back(circle);
+      } catch (const std::exception & ex) {
+        RCLCPP_ERROR(get_logger(), "Can not add circle: %s", ex.what());
+        response->success = false;
+      }
+    }
+  }
+
+  switchMapUpdate();
+}
+
+void VectorObjectServer::getShapesCallback(
+  const std::shared_ptr<rmw_request_id_t>/*request_header*/,
+  const std::shared_ptr<nav2_msgs::srv::GetShapes::Request>/*request*/,
+  std::shared_ptr<nav2_msgs::srv::GetShapes::Response> response)
+{
+  std::shared_ptr<Polygon> polygon;
+  std::shared_ptr<Circle> circle;
+
+  for (auto shape : shapes_) {
+    switch (shape->getType()) {
+      case POLYGON:
+        polygon = std::static_pointer_cast<Polygon>(shape);
+        response->polygons.push_back(*(polygon->getParams()));
+        break;
+      case CIRCLE:
+        circle = std::static_pointer_cast<Circle>(shape);
+        response->circles.push_back(*(circle->getParams()));
+        break;
+      default:
+        RCLCPP_WARN(get_logger(), "Unknown shape type (UUID: %s)", shape->getUUID().c_str());
+    }
+  }
+}
+
+void VectorObjectServer::removeShapesCallback(
+  const std::shared_ptr<rmw_request_id_t>/*request_header*/,
+  const std::shared_ptr<nav2_msgs::srv::RemoveShapes::Request> request,
+  std::shared_ptr<nav2_msgs::srv::RemoveShapes::Response> response)
+{
+  // Initialize result with true. If one of the required vector object was not found,
+  // set it to false.
+  response->success = true;
+
+  for (auto req_uuid : request->uuids) {
+    bool found = false;
+
+    for (
+      std::vector<std::shared_ptr<Shape>>::iterator it = shapes_.begin();
+      it != shapes_.end();
+      it++)
+    {
+      if ((*it)->isUUID(req_uuid.uuid.data())) {
+        // Polygon with given UUID was found: remove it
+        (*it).reset();
+        shapes_.erase(it);
+        found = true;
+
+        break;
+      }
+    }
+
+    if (!found) {
+      // Required vector object was not found
+      char uuid_str[37];
+      uuid_unparse(req_uuid.uuid.data(), uuid_str);
+      RCLCPP_ERROR(get_logger(), "Can not find shape to remove with UUID: %s", uuid_str);
+      response->success = false;
+    }
+  }
+
+  switchMapUpdate();
+}
+
+bool VectorObjectServer::getROSParameters()
+{
+  // Main ROS-parameters
+  map_topic_ = getROSParameter(
+    shared_from_this(), "map_topic", "vo_map").as_string();
+  frame_id_ = getROSParameter(
+    shared_from_this(), "frame_id", "map").as_string();
+  resolution_ = getROSParameter(
+    shared_from_this(), "resolution", 0.05).as_double();
+  default_value_ = getROSParameter(
+    shared_from_this(), "default_value", nav2_util::OCC_GRID_UNKNOWN).as_int();
+  overlay_type_ = static_cast<OverlayType>(getROSParameter(
+      shared_from_this(), "overlay_type", static_cast<int>(OverlayType::OVERLAY_SEQ)).as_int());
+  update_frequency_ = getROSParameter(
```

Comment:
Is there any logic to skip updates if all of the objects are in the map frame so its static and no new shapes were added?

Reply:
Yes, it is switched be the following comparison method:
```
bool VectorObjectServer::isMapUpdate()
{
  for (auto shape : shapes_) {
    if (shape->getFrameID() != global_frame_id_ && !shape->getFrameID().empty()) {
      return true;
    }
  }

  return false;
}
```

If shape's `frame_id` differs from `global_frame_id_` and shape has it (empty frame id for the shape is treated to be equal to global frameID by-default), we are publishing map dynamically. Otherwise - statically once (per each shape update/remove call).